### DISCUSSION TITLE
Serverless v1 compatibility with AWS Lambda states

### DISCUSSION THREAD

**Discussion by bryanowen7** [serverless/serverless#10277] [Q&A] [answered]
Now that AWS Lambda states have been introduced ([more information here](https://aws.amazon.com/blogs/compute/coming-soon-expansion-of-aws-lambda-states-to-all-functions/)), Serverless v1 (1.83.3) has become unusable.

Whenever I try to deploy a function, I get this error mesage: _"The operation cannot be performed at this time. An update is in progress for resource :function arn:"_. It seems like serverless is not waiting for the state to change. This hasn't been implemented for obvious reasons: Serverless v1 hasn't been updated for a year, unless I'm missing something.
I've been able to avoid this change by opting out of the change by changing the description of my functions. However, on December 6th, this feature will no longer be optional.

I have some important microservices running on Serverless v1 which I need to keep updating. Without a fix, I won't be able to deploy any more changes to my functions.
Please note I am completely aware that Serverless v1 has been deprecated, and that there are multiple reasons I should migrate to the latest stable version as soon as possible. However, I need to delay the transition for a couple of months to focus on other urgent projects. I also use the plugin [Serverless Alias](https://github.com/serverless-heaven/serverless-aws-alias) which has been long abandoned and only works with Serverless v1. This is the main reason I can't afford to update right now. I would need to update multiple systems using these microservices and that would be very time-consuming.

I'm wondering if there's something I can do (official or not) to keep using Serverless v1. I will update to Serverless v2 when I get the chance, but for the time being, I would like to keep using that version for at least the rest of the year.

Thank you in advance.

**Comment by pgrzesik** [comment]
Hello @bryanowen7 👋 Thanks a lot for your report. Could you please provide more details on when does that happen for you, ideally with detailed logs with `SLS_DEBUG=*` set. We'd like to first understand what might be causing this issue for you.

**Reply by bryanowen7** [reply]
It happens basically every time I try to deploy a function.
The command I executed for this example is: `sls deploy -s dev --force -f api-validate-dte`. The _-s dev --force_ arguments are part of the serverless-aws-alias plugin.
Logs: [log.txt](https://github.com/serverless/serverless/files/7605386/deploy-log.txt)

When I go to the AWS Console to check the code of the function with the web editor, my changes are uploaded, but no new version is created and changes remain in the $LATEST version only.

Once I update the description on the function to add "_aws:states:opt-out_", and try to deploy the function, it works as intended again.

**Reply by whatwg6** [reply]
Maybe remove --functions params solve this problem。

https://github.com/serverless/serverless/discussions/10271#discussioncomment-1702710

**Reply by bryanowen7** [reply]
> Maybe remove --functions params solve this problem。
> 
> [#10271 (comment)](https://github.com/serverless/serverless/discussions/10271#discussioncomment-1702710)

Yes, deploying the whole service seems to allow me to deploy, however, this is not a good solution for stacks with dozens of functions like mine
At the end I decided to bite the bullet and update all my services to serverless v2

**Reply by pgrzesik** [reply]
Hello @bryanowen7 👋 How does it behave for you on `v2` version?

**Reply by bryanowen7** [reply]
> Hello @bryanowen7 👋 How does it behave for you on `v2` version?

Using the latest version of v2 works fine for me

**Reply by pgrzesik** [reply]
Great to hear that 👍

**Comment by beppojs** [comment]
It seems that this is not only a problem for serverless v1, I am facing the same problem for serverless v2 (2.5.0), see #10271

**Comment by pgrzesik** [comment, marked as answer]
Hello @bryanowen7 - we've investigated more and that issue will be fully addressed in `v2` branch. We do not plan any backports for `v1` as we're currently focusing on `v3` release and `v1` has been marked as deprecated. Please note that only `deploy function` is affected and all other functionalities should work without issues.

**Reply by blibla5** [reply]
Do you know, when this issue will be fixed?

**Reply by pgrzesik** [reply]
It has been addressed already - what version of the Framework are you currently using? @blibla5

**Reply by blibla5** [reply]
I am using v2.68.0 @pgrzesik

**Reply by pgrzesik** [reply]
What exact issue are you still experiencing @blibla5?

**Reply by topisark** [reply]
I can also confirm that on 2.68.0 deploying a single function behaves strangely:

```
Serverless: Packaging function: calls...
Serverless: Uploading function: calls (639 kB)...
Serverless: Successfully deployed function: calls
Serverless: Retrying configuration update for function: calls. Reason: The operation cannot be performed at this time. An update is in progress for resource: arn:aws:lambda:eu-central-xxxx
Serverless: Retrying configuration update for function: calls. Reason: The operation cannot be performed at this time. An update is in progress for resource: arn:aws:lambda:eu-central-xxxx
(this repeats about 20 times)
Serverless: Successfully updated function: calls
Serverless: Retrying ensure function state for function: calls.
Serverless: Retrying ensure function state for function: calls.
(this repeats about 20 times)
````

It does eventually go through.

**Reply by pgrzesik** [reply]
@topisark This is something outside of Framework's control and happens on AWS side, do you use VPC with your functions by any chance?

**Reply by topisark** [reply]
The function is inside VPC, yes.

**Reply by pgrzesik** [reply]
I've been observing the same problem for functions within VPC, while the problem isn't reproducible for me outside of VPC. It seems like something changed on AWS side about updates of VPC-based functions which is slowing down the whole cycle.

**Reply by blibla5** [reply]
I face the exact same problem as topisark describes.

**Reply by pgrzesik** [reply]
@blibla5 Are you using Lambdas in VPC as well?

**Reply by blibla5** [reply]
@pgrzesik No Lambdas are not in the VPC. But I only get the error once, if the Lambda description doesn't contain 'aws:states:opt-out'.
Is the bug fixed in v2.68.0?

**Reply by pgrzesik** [reply]
Yes, it is fixed. The retry notice that you're seeing is expected, we have to wait for an update because otherwise the update will crash., it's outside of Framework control

**Reply by fernandofleury** [reply]
I'm actually getting the `ResourceConflictException: The operation cannot be performed at this time. An update is in progress for resource arn:aws:lambda` after upgrading to `2.68.0`. What else do we need to change in the config to have it working now? The state documentation is a bit abstract.

Here's what I currently have that was working prior to it (at `.serverless/template.name.json`):

```json
{
  "name": "name",
  "arn": "arn",
  "service": [
    "edgelambda.amazonaws.com",
    "lambda.amazonaws.com"
  ],
  "policy": {
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Resource": "*",
        "Action": [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents"
        ]
      },
      {
        "Effect": "Allow",
        "Resource": "arn:aws:s3:::scope/*",
        "Action": [
          "s3:GetObject",
          "s3:PutObject"
        ]
      }
    ]
  },
  "region": "region"
}
```

**Reply by pgrzesik** [reply]
Hello @fernandofleury - what version of the Framework are you using? Would you be able to provide a small reproducible example for this or a sample config?

**Reply by fernandofleury** [reply]
@pgrzesik Currently running on `serverless@2.68.0`. Is there anything else we need to change in order to enable the state waiting? Or anything specific I should be looking at the config? Otherwise, I'll see what I can do about the repro

**Reply by fernandofleury** [reply]
we were able to sort it out by upgrading an external component (called `serverless-next`). That also had to be updated :)

**Reply by pgrzesik** [reply]
Thanks for an update @fernandofleury 🙇
